state/watcher.go

```diff
@@ -3709,40 +3709,143 @@ func (w *containerAddressesWatcher) loop() error {
 }
 
 func newSettingsHashWatcher(st *State, localID string) StringsWatcher {
-	w := &settingsHashWatcher{
+	docID := st.docID(localID)
+	w := &hashWatcher{
 		commonWatcher: newCommonWatcher(st),
 		out:           make(chan []string),
-		id:            st.docID(localID),
-		name:          localID,
+		collection:    settingsC,
+		id:            docID,
+		hash: func() (string, error) {
+			return hashSettings(st.db(), docID, localID)
+		},
 	}
-	w.tomb.Go(func() error {
-		defer close(w.out)
-		return w.loop()
-	})
+	w.start()
+	return w
+}
+
+func hashSettings(db Database, id string, name string) (string, error) {
+	settings, closer := db.GetCollection(settingsC)
+	defer closer()
+	var doc settingsDoc
+	if err := settings.FindId(id).One(&doc); err == mgo.ErrNotFound {
+		return "", nil
+	} else if err != nil {
+		return "", errors.Trace(err)
+	}
+	// Ensure elements are in a consistent order. If any are maps,
+	// replace them with the equivalent sorted bson.Ds.
+	items := toSortedBsonD(doc.Settings)
+	data, err := bson.Marshal(items)
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+	hash := sha256.New()
+	hash.Write([]byte(name))
+	hash.Write(data)
+	return fmt.Sprintf("%x", hash.Sum(nil)), nil
+}
+
+// WatchAddressesHash returns a StringsWatcher that emits the hash of
+// the machine's (sorted) addresses whenever they change.
+func (m *Machine) WatchAddressesHash() StringsWatcher {
+	w := &hashWatcher{
+		commonWatcher: newCommonWatcher(m.st),
+		out:           make(chan []string),
+		collection:    machinesC,
+		id:            m.doc.DocID,
+		hash: func() (string, error) {
+			if err := m.Refresh(); err != nil {
+				return "", errors.Trace(err)
+			}
+			addresses := m.Addresses()
+			sort.Slice(addresses, func(i, j int) bool {
+				// Addresses guarantees that each value will only be
+				// returned once - addresses from provider take
+				// precedence over those from the machine.
+				return addresses[i].Value < addresses[j].Value
+			})
+			hash := sha256.New()
+			for _, address := range addresses {
+				hash.Write([]byte(address.Value))
+				hash.Write([]byte(address.Type))
+				hash.Write([]byte(address.Scope))
```

Comment:
scope is included in the hash yet in some tests where scope is change but address is the same the watcher doesn't fire?